### ISSUE TITLE
Use constructors when instantiating ICommand and IEvent

### ISSUE THREAD

**Issue by isaacabraham** [Particular/NServiceBus#6061]
If you create a Record in F# (and I assume C#9 is the same, but might be wrong), you essentially end up with a class that has a public constructor taking in every value, with getter-only props:

```fsharp
type PlaceOrder =
    { OrderId: string }
    interface ICommand
```

would boil down to something like

```csharp
public class PlaceOrder :ICommand {
    string OrderId {get; private set; }
    public PlaceOrder (string orderId)  {
        OrderId = orderId;
    }
}
```

This doesn't work with NSB - `OrderId` is left blank when rehydrating one, with no error giving - the field is simply null.

To work around this, you have to decorate the record with `[<CLIMutable>]`, which tells F# to create the class with a public setter as well. This works, but it's not obvious to the developer and not doing it doesn't lead to any eager exceptions / errors.

**Comment by timbussmann** [comment]
that behavior mostly seems to rely on the serializer and it's capability to deal with that. Is this the behavior you're seeing on both XML and JSON serializers? How would Newtonsoft.Json and System.Json deal when deserializing such types?

**Comment by fjmorel** [comment]
In C#, I have had no issues with records using `Newtonsoft.Json`. I had issues using records with another library that used `SimpleJson` instead.

**Comment by timbussmann** [comment]
@fjmorel we use SimpleJson internally in NServiceBus and some downstream packages, but for message deserialization, it should use the configured serializer (typically newtonsoft.json or XML). Are you experiencing the same issue? Do you have a small repro sample to look at maybe?

**Comment by fjmorel** [comment]
We use the `Newtonsoft` serializer, so we've had no issues with records.

I was just trying to point out that it was a problem with `SimpleJson`, but that other serializers likely don't have the issue.

**Comment by timbussmann** [comment]
> I was just trying to point out that it was a problem with SimpleJson, but that other serializers likely don't have the issue.

👍 we already do have an issue to replace SimpleJson https://github.com/Particular/NServiceBus/issues/6227, which might happen as soon as we can use System.Text.Json on every build target. I can't immediately come up with a case where this issue can affect users, but if you are aware of something, please let us know.

**Comment by sylvaingirardbe** [comment]
> 👍 we already do have an issue to replace SimpleJson #6227, which might happen as soon as we can use System.Text.Json on every build target. I can't immediately come up with a case where this issue can affect users, but if you are aware of something, please let us know.

We're running into more or less the same issue using immutable records in sagadata. Events work perfectly fine using immutable records as long as configure the serialization to use Newtonsoft.

AFAIK, should you replace SimpleJson with Newtonsoft, you wouldn't need to wait for System.Text.Json on every build target.

**Comment by sylvaingirardbe** [comment]
> What kind of errors do you run into exactly? Does this cause issues only when deserializing a saga? can you share some example of such a saga data?
Correct, when deserializing the saga we get a null reference on line 1435 in SimpleJson (constructor not found).

Take this record type of which you have a proeprty in SagaData, for instance. This will work:
```
    public record VehicleInfo
    {
        public VehicleInfo()
        { }
        
        public VehicleInfo(
            VehicleType type, 
            string identifier, 
            string nationality, 
            IEnumerable<string> virtualStorageUnitCodes
        )
        {
            Type = type;
            Identifier = identifier;
            Nationality = nationality;
            VirtualStorageUnitCodes = virtualStorageUnitCodes;
        }
        
        public VehicleType Type { get; set; }
        public string Identifier { get; set; }
        public string Nationality { get; set; }
        public IEnumerable<string> VirtualStorageUnitCodes { get; set; }
    }
```
If you make the type immutable, or in this case just remove the default ctor, it'll fail with a null reference. Deserializing this type using Newtonsoft or System.Json works just fine.

> A major reason to use SimpleJson is the ability to ship it as part of NServiceBus without the need of an external package dependency. NServiceBus versions are commonly used over many years and managing an external dependency is very tricky to get right (and heavily depends on the external package's maintainers approach to versioning and compatibility).

OK, got it.

**Comment by timbussmann** [comment]
> If you make the type immutable, or in this case just remove the default ctor, it'll fail with a null reference. Deserializing this type using Newtonsoft or System.Json works just fine.

thanks for the example. I'll forward this internally to look into. Both Newtonsoft.Json and System.Text.Json seem to be fine with readonly structs while SimpleJson hasn't been designed for these new language features. SimpleJson requires an empty ctor, but it seems to work fine with `init` properties, so you could still make your record effectively read-only by changing it to something like this:

```
public record VehicleInfo
{
    public VehicleInfo()
    {
    }

    public VehicleInfo(
        VehicleType type,
        string identifier,
        string nationality,
        IEnumerable<string> virtualStorageUnitCodes
    )
    {
        Type = type;
        Identifier = identifier;
        Nationality = nationality;
        VirtualStorageUnitCodes = virtualStorageUnitCodes;
    }

    public VehicleType Type { get; init; }
    public string Identifier { get; init; }
    public string Nationality { get; init; }
    public IEnumerable<string> VirtualStorageUnitCodes { get; init; }
}
```

**Comment by sylvaingirardbe** [comment]
> thanks for the example. I'll forward this internally to look into. Both Newtonsoft.Json and System.Text.Json seem to be fine with readonly structs while SimpleJson hasn't been designed for these new language features. SimpleJson requires an empty ctor, but it seems to work fine with `init` properties, so you could still make your record effectively read-only by changing it to something like this:

Thank for the feedback. We didn't try that combination. In any case this forces us to type out the entire property definition hoopla, instead of it basically being a one-line record definition like so
```
public record VehicleInfo(VehicleType Type, string Identifier, string Nationality, IEnumerable<string> VirtualStorageUnitCodes);
```

We a lazy bunch :)

**Comment by timbussmann** [comment]
yeah, the one-line positional record definition doesn't provide an empty ctor (and it seems it can't be configured to get one), so that's a bit unfortunate. This should be enabled once #6227 has been completed though.

**Comment by timbussmann** [comment]
btw. @sylvaingirardbe what transport and persistence are you using?

**Comment by sylvaingirardbe** [comment]
> btw. @sylvaingirardbe what transport and persistence are you using?

Differs per environment. Transport is either RabbitMQ or ASB and persistence SQL Server or Azure Table.
